### PR TITLE
fix(api,shared-data): Lowercase labware names and make categories consistent

### LINKED ISSUE #3231: Labware Library: Shared data cleanup
## overview

In order to accurately display labware data to the user, we need to clean up a few things in the v2 labware definitions in `shared-data/definitions2`.

### existing definitions

- [ ] Make all filenames lowercase
- [ ] `camelCase` all display categories
- [ ] Ensure all categories are one of:
    - `tipRack`
    - `tubeRack`
    - `trough`
    - `trash`
    - `wellPlate`
    - `other`
- [ ] Add `"brand": {"brand": "generic"}` for generics

### generator

- [ ] Ensure output filenames are lowercase
- [ ] Add categories and filename requirements to schema
- [ ] Move types to common location
- [ ] Ensure `brand` always exists

### PR DESCRIPTION
**Pull request by mcous**
## overview

In the process of implementing #3075, a few things came up:

- A number of data inconsistencies / schema + typedef differences
- A need to use typedefs that only existed in the labware generator

Closes #3231

## changelog

All changes were made to existing definitions, the labware generator, and the labware schema

- Moved labware definition typedef to `shared-data/js/types.js`
- Made all labware filenames + load names lowercase
    - Updated PAPIv2's load method to always lowercase the input name
- `camelCase`d all display categories and restricted them to:
    - `tipRack`
    - `tubeRack`
    - `trough`
    - `trash`
    - `wellPlate`
    - `other`
- Instead of omitting `brand`, generic labware now have `"brand": {"brand": "generic"}`
- Corrected various bugs / misalignments in the schema + typedef
    - Schema: `x`, `y`, and `z` were not marked as required for `wells`
    - Schema: `brand.brand` was not marked as required if `brand` was present
    - Typedef: `parameters.loadName` was not marked as required

## review requests

This PR affects both the labware generator and PAPIv2, so if you can:

- [ ] Run a regular labware generation
- [ ] Run an irregular labware generation
- [ ] Labware can be loaded in PAPIv2 with a lowercase name in the load statement
- [ ] Labware can be loaded in PAPIv2 with a mixed-case name in the load statement

Also please check out the documentation updates!

### CODE REVIEW

File: api/docs/source/new_protocol_api.rst [hunk 1]
```diff
@@ -127,49 +127,49 @@ This table lists the names of valid labwares that can be loaded with :py:meth:`.
...
+| opentrons_96_tiprack_1000_ul              | tiprack-1000ul                         |
 +-------------------------------------------+----------------------------------------+
-| opentrons_96_tiprack_10_uL                | tiprack-10ul                           |
+| opentrons_96_tiprack_10_ul                | tiprack-10ul                           |
 +-------------------------------------------+----------------------------------------+
-| opentrons_96_tiprack_300_uL               | opentrons-tiprack-300ul                |
+| opentrons_96_tiprack_300_ul               | opentrons-tiprack-300ul                |
 +-------------------------------------------+----------------------------------------+
-| usa_scientific_12_trough_22_mL            | usa_scientific_12_trough_22_mL         |
+| usa_scientific_12_trough_22_ml            | usa_scientific_12_trough_22_ml         |
```
Comment (Laura-Danielle) [inline api/docs/source/new_protocol_api.rst]: Not sure if this is meant to be a mapping in the docs, but the equivalent definition in V1 of `usa_scientific_12_trough_22ml` is actually `trough-12row`

File: shared-data/js/labwareTools/index.js [hunk 2]
```diff
@@ -141,10 +142,7 @@ export function _generateIrregularLoadName (args: {
     return `${numWells}x${well[gridIdx].totalLiquidVolume}_${units}`
   })
 
-  return [brand, wellComboArray.join('_'), displayCategory]
-    .join('_')
-    .replace(' ', '_')
```
Comment (mcous) [inline shared-data/js/labwareTools/index.js]: Note: this call to `replace` only replaced the _first_ space with an underscore rather than all spaces

File: shared-data/js/labwareTools/index.js [hunk 3]
```diff
@@ -204,6 +202,17 @@ function ensureBrand (brand?: Brand): Brand {
...
 
+// joins the input array with _ to create a name, making sure to lowercase the
+// result and remove all invalid characters (allowed characters: [a-z0-9_.])
+function createName (
+  fragments: Array<string | number | Array<string | number>>
+): string {
+  return flatten(fragments)
+    .join('_')
+    .toLowerCase()
+    .replace(/[^a-z0-9_.]/g, '')
```
Comment (mcous) [inline shared-data/js/labwareTools/index.js]: I decided to remove invalid characters rather than replace them with `_` by going off our existing Bio-Rad labware. The brand name is "Bio-Rad", and our existing labware defs are named `biorad_...`, not `bio_rad` nor `bio-rad_...`. Let me know if this change is not what we want

Comment (Laura-Danielle) [reply shared-data/js/labwareTools/index.js]: I think that should be fine for now. I can't think of any edge cases right now where this would result in an incorrect/confusing name at the moment.

File: api/tests/opentrons/protocol_api/test_module_context.py [hunk 4]
```diff
@@ -128,11 +128,11 @@ def test_thermocycler_temp(loop):
 
 def test_module_load_labware(loop):
     ctx = papi.ProtocolContext(loop)
-    labware_name = 'generic_96_wellPlate_380_uL'
+    labware_name = 'generic_96_wellplate_380_ul'
     labware_def = json.loads(
         pkgutil.get_data('opentrons',
                          'shared_data/definitions2/{}.json'.format(
-                             labware_name)))
+                             labware_name.lower())))
```
Comment (IanLondon) [inline api/tests/opentrons/protocol_api/test_module_context.py]: why's it lowercasing here and a bunch of other places in the tests?

File: labware-designer/README.md [hunk 5]
```diff
@@ -39,57 +40,79 @@ OR
...
-by typing `sharedData.createRegularLabware(input)` or `sharedData.createIrregularLabware(input)`
+by typing `sharedData.createRegularLabware(options)` or `sharedData.createIrregularLabware(options)`
 
 ### What data is needed?
-To build a *regular* labware, the `input` object should have the following shape:
 
+To build a _regular_ labware, the `options` object should have the following shape:
+
+```js
+const options = {
```
Comment (IanLondon) [inline labware-designer/README.md]: this format is a bit confusing b/c it looks kinda like code you can copy-paste into the console especially with the call at the end, but really it's using Flow as shorthand to describe the shape. Just taking out the `const` and the call to sharedData.createRegularLabware might make it look more like a type def description

File: labware-designer/README.md [hunk 6]
```diff
@@ -201,319 +246,289 @@ See diagram below:
 
 Given this sequence of commands in the console:
```
Comment (IanLondon) [inline labware-designer/README.md]: This simple copy-and-pastable example might be a better introduction to put at the top, rather than starting with the detailed definition. Matter of taste though!

File: api/tests/opentrons/protocol_api/test_module_context.py (hunk 4 above)
Comment (mcous) [reply api/tests/opentrons/protocol_api/test_module_context.py]: Ah that was a leftover from when I thought the necessary changes were less extensive. I got a little paranoid because the tests that were failing were only failing on Linux, not macOS (case-insensitive filesystem and all). Will remove

I only count two times in the tests, though; both in this file. Am I missing some? There's one more in the source that is intentional and necessary to avoid breaking changes

File: shared-data/js/labwareTools/__tests__/createIrregularLabware.test.js [hunk 7]
```diff
@@ -49,19 +57,35 @@ describe('test helper functions', () => {
...
     const expectedX1 = [1, 11, 21, 31, 41]
     const expectedY1 = [0.5]
-    const expectedX2 = [ 1, 15, 29, 43, 57 ]
+    const expectedX2 = [1, 15, 29, 43, 57]
     const expectedY2 = [0.5]
 
     range(grid.column).forEach(colIdx => {
       range(grid.row).forEach(rowIdx => {
-        const well1 = _calculateWellCoord(rowIdx, colIdx, spacing[0], offset, well[0])
+        const well1 = _calculateWellCoord(
```
Comment (IanLondon) [inline shared-data/js/labwareTools/__tests__/createIrregularLabware.test.js]: aside: prettier is great and we should use it more widely IMO, but it generates a lot of noise in the diff when you apply it to "un-pretty" files like this one :frowning:
